### PR TITLE
Plugins: Check registration keys on site before automatically setting up purchased plugins

### LINKED ISSUE #9320: Plugins: Overwrites existing service keys when setting up a new plan
When going through the plugin setup process after buying a plan for a Jetpack site, we currently overwrite the API keys for Polldaddy and Akismet, if the user already has these config'd. This can lead to apparent loss of existing data (polls or stats), because the new keys have no history associated.

We've already addressed this in VaultPress by not registering the new key, plus the process of generating the key _should_ fetch same key that the site already uses. Polldaddy generates a new key for this registration process, even if the site is already registered, and I don't think we can get around that. Akismet is only overwritten if the Jetpack site is connected to a different wp.com account than the existing Akismet key.

We could try to check if a plugin is already registered when trying to set it up, and notify the user of the existing connection - and have them contact support to untangle their plans? (so they aren't paying for multiple of the same product) I'm not sure how else we can help users to help themselves with this.

@johnHackworth @beaulebens @rickybanister (to think through what we should show users)

### PR DESCRIPTION
**Pull request by ryelle**
This fixes #9320: Plugin registration options are overwritten if the user already has, for example, Akismet registered under a different account.

This PR adds a check for the current registration option for each plugin before trying to save the new registration key. It will detect if the plugin is already registered with the given key, and do nothing (considered a successful setup), or detect if the plugin is already configured with a different key. If it's already registered with a different key, we'll show a little info notice. Since technically everything should be working, I don't think an error state makes sense, so even in this case it ends in "success"

To test:

- Set up Akismet under one wp.com account
- Connect Jetpack on that site to a _different_ wp.com account, and buy a plan
- Go through the autoconfig flow (should be prompted after buying plan, or you can go directly to `/plugins/setup/$site`)
- The plugins will run through installing, and once everything's finished, you should see a blue notice under Akismet, saying "This plugin is already registered with another plan."

screen shot 2016-11-29 at 2 59 12 pm

cc @johnHackworth @rickybanister

### CODE REVIEW

**Comment by johnHackworth** [comment]
Just partially related with this PR purpose... While trying to test it, I just got this error notice with a malformed <a> link:

image

File: client/my-sites/plugins/jetpack-plugins-setup/index.jsx [hunk 1]
```diff
@@ -398,7 +400,7 @@ const PlansSetup = React.createClass( {
 		}
 
 		const pluginsWithErrors = filter( this.props.plugins, ( item ) => {
-			return ( item.error !== null );
+			return ( item.error !== null ) && ( item.error.code !== 'already_registered' );
```
Comment (gwwar) [inline client/my-sites/plugins/jetpack-plugins-setup/index.jsx]: Optional, but lodash get can make this a bit more readable:
```
const errorCode = get( item, 'error.code', null );
return  errorCode && errorCode !== 'already_registered';
```

File: client/my-sites/plugins/jetpack-plugins-setup/index.jsx [hunk 2]
```diff
@@ -299,6 +299,11 @@ const PlansSetup = React.createClass( {
 					{ translate( 'Manual Installation' ) }
 				</NoticeAction>
 			);
+			if ( plugin.error.code === 'already_registered' ) {
+				statusProps.status = 'is-info';
+				statusProps.text = translate( 'This plugin is already registered with another plan.' );
+				delete statusProps.children;
+			}
```
Comment (gwwar) [inline client/my-sites/plugins/jetpack-plugins-setup/index.jsx]: We can leave this for another PR but this logic is a bit difficult for me to follow.

What do folks think of breaking this into two functions `getPluginErrorStatus` and `getPluginStatus` where we return jsx within each switch case instead of falling through?

File: client/state/plugins/premium/actions.js [hunk 3]
```diff
@@ -209,27 +209,59 @@ function configure( site, plugin, dispatch ) {
...
 		} );
 	}
-	site.setOption( { option_name: option, option_value: optionValue }, ( error, data ) => {
-		if ( ( 'vaultpress' === plugin.slug ) && versionCompare( plugin.version, '1.8.3', '>' ) ) {
-			const response = JSON.parse( data.option_value );
-			if ( 'response' === response.action && 'broken' === response.status ) {
-				error = new Error( response.error );
-				error.name = 'RegisterError';
+
+	const saveOption = () => {
```
Comment (gwwar) [inline client/state/plugins/premium/actions.js]: Unless I'm misreading this, it looks like we can define this outside of this function with: 
```
saveOption( site, plugin, dispatch ) {
    return return site.setOption( { option_name: option, option_value: optionValue }, ( error, data ) => { ... } );
);
}
```

File: client/state/plugins/premium/actions.js [hunk 4]
```diff
@@ -209,27 +209,59 @@ function configure( site, plugin, dispatch ) {
...
+		} );
+	};
+
+	// We don't need to check for VaultPress
+	if ( 'vaultpress' === plugin.slug ) {
+		return saveOption();
+	}
+
+	return site.getOption( { option_name: option }, ( getError, getData ) => {
+		if ( getData.option_value && getData.option_value === optionValue ) {
```
Comment (gwwar) [inline client/state/plugins/premium/actions.js]: Maybe
```
if ( get( getData, 'option_value' ) === optionValue ) {
```

**Comment by gwwar** [comment]
The expected notice shows up 👍  I left a few notes, but should be good to 🚢 if no one finds any blockers.

Note that I think my Jetpack site is a bit funky (had issues with VaultPress before) when adding/removing plans.
screen shot 2016-12-02 at 4 48 26 pm

**Comment by ryelle** [comment]
I switched to lodash's `get` where suggested. I started to refactor `getPluginStatus`, but I also think that should wait for a second PR. Still good to 🚢 ?

**Comment by rickybanister** [comment]
I'm somewhat confused by the 'this plugin is already configured with another plan' language. We don't want to overwrite their current configuration, but we also don't want them paying twice. 

I think we can avoid this situation better further upstream by making the plans page more aware of existing old-style plans, but if they do get to this point perhaps we should be slightly more proactive.

Can we have that notice say something more like `Akismet was already configured    |   Manage Plans`? 

@beaulebens @richardmuscat @annezazuu for further thoughts here, basically I just want to make sure we're creating a nice path forward for consolidation, rather than furthering any confusion.

**Comment by gwwar** [comment]
Code looks good to me 👍 Should be fine to 🚢 when folks are happy with the copy. Editorial might be able to help too.

**Comment by ryelle** [comment]
I'm going to merge this as-is, because it's a pretty rare case (except for a12s) but it would be nice to have a fix for people who do run into it, and we can continue the copy discussion in the PR I'll open immediately after this :)
